Knowledge tuning push tests

andrewdonheiser opened the pull request:
Feature: Add push tests for knowledge-tuning
# Usecase - iLab Knowledge Pipeline

## Description

This PR adds a comprehensive push test suite for the `examples/knowledge-tuning/` directory, including 43 tests that validate notebook structure, imports, utility functions, environment configuration, and virtual environment builds. The implementation features fast execution (~18 seconds), parallel testing, and complete documentation following the open AGENTS.md standard (https://agents.md/).

**Motivation**: Ensure notebooks and code in the knowledge-tuning directory meet quality standards before merge, with automated validation that runs on every commit and pull request.

**Key Features**:
- 43 comprehensive tests covering notebook structure, imports, utilities, environment, and venv builds
- Fast execution (~18 seconds) using parallel testing with pytest-xdist
- Complete mock infrastructure for transformers, GPU operations, and external APIs
- Automatic test execution on all commits and pull requests via GitHub Actions
- Comprehensive documentation following AGENTS.md standard (source of truth for agent instructions)

**Test Coverage**:
- ✅ 39 tests passing
- ⏭️ 2 tests skipped (GPU dependencies on non-GPU systems - expected)
- ❌ 2 tests failing (notebook cleanup needed - existing issues in Base_Model_Evaluation.ipynb)

Fixes # (issue)

## Related Tickets

Related jira tickets here:

- [RHAIENG-1623](https://issues.redhat.com/browse/RHAIENG-1623)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

**Test Suite Added**:
- [x] Notebook Structure Tests - Validates metadata compliance, no execution counts, no stored outputs
- [x] Import Validation Tests - Syntax checking of import statements without execution
- [x] Utility Function Tests - Tests utility functions with mocked dependencies
- [x] Environment Tests - Validates pyproject.toml files and dependency configuration
- [x] Venv Build Tests - Validates virtual environment builds for all 6 steps

**Test Execution**:h
# Install dependencies
cd tests/knowledge-tuning
pip install -e .

# Run all tests (includes parallel execution and reports)
python tests/knowledge-tuning/run_push_tests.py

# Or run with pytest directly
pytest tests/knowledge-tuning/ -v**Test Results**: Automatically generated in `tests/knowledge-tuning/test-results/`:
- JUnit XML report for CI/CD integration
- HTML report for human review

**Test Configuration**:
- Python version: 3.12
- pytest version: >=7.4.0
- pytest-xdist version: >=3.3.0 (parallel execution)
- pytest-html version: >=4.0.0 (test reports)
- nbformat version: >=5.9.0
- polars version: >=0.20.0
- jsonschema version: >=4.19.0

**CI/CD**: Tests run automatically on all commits and pull requests via GitHub Actions workflow (`.github/workflows/knowledge-tuning-push-tests.yml`)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

## Files Added

**Test Infrastructure** (15 files, 1,626 insertions):
- `tests/knowledge-tuning/__init__.py` - Package initialization
- `tests/knowledge-tuning/conftest.py` - Pytest fixtures and shared utilities
- `tests/knowledge-tuning/test_notebook_structure.py` - Notebook structure validation (227 lines)
- `tests/knowledge-tuning/test_notebook_imports.py` - Import validation (183 lines)
- `tests/knowledge-tuning/test_utils.py` - Utility function testing (257 lines)
- `tests/knowledge-tuning/test_environment.py` - Environment configuration testing (98 lines)
- `tests/knowledge-tuning/test_venv_build.py` - Virtual environment build validation (283 lines)
- `tests/knowledge-tuning/run_push_tests.py` - Test runner script with parallel execution
- `tests/knowledge-tuning/pyproject.toml` - Test dependencies configuration
- `tests/knowledge-tuning/mocks/__init__.py` - Mock package initialization
- `tests/knowledge-tuning/mocks/transformers_mock.py` - Mock implementations for HuggingFace transformers
- `.github/workflows/knowledge-tuning-push-tests.yml` - GitHub Actions workflow
- `tests/knowledge-tuning/AGENTS.md` - Source of truth for agent instructions (235 lines)
- `tests/knowledge-tuning/README.md` - Human-focused guide (99 lines)
- Updated `.gitignore` - Added test results and cache file exclusions

andrewdonheiser committed:
feat: Add push tests for knowledge-tuning with AGENTS.md documentation

Add comprehensive push test suite for knowledge-tuning directory with fast
execution, parallel testing, and complete documentation following AGENTS.md
standard.

Test Infrastructure:
- Complete test suite (43 tests) validating notebooks, imports, utilities,
  environment, and venv builds
- Parallel execution with pytest-xdist (~18 seconds total runtime)
- Test result generation (JUnit XML, HTML reports)
- Mock infrastructure for transformers, GPU operations, and external APIs
- GitHub Actions workflow for automatic execution on push

Test Coverage:
- Notebook structure validation (metadata, execution counts, outputs)
- Import statement validation (syntax checking without execution)
- Utility function testing with mocked dependencies
- pyproject.toml validation and dependency resolution
- Virtual environment build validation (6 venvs tested)

Documentation:
- AGENTS.md: Source of truth for agent instructions (converted from .cursorrules)
- README.md: Human-focused guide with setup and usage instructions
- Both files reference AGENTS.md as authoritative source

Configuration:
- Updated .gitignore to exclude test results and cache files
- Test dependencies in pyproject.toml (pytest, pytest-xdist, pytest-html, etc.)
- Runner script with parallel execution and report generation

All tests follow .cursorrules requirements: no notebook execution, all
external services mocked, fast execution, CI consistency.

andrewdonheiser committed:
ci: Run push tests on pull requests

Update workflow to trigger on pull requests in addition to pushes to main,
ensuring tests run when PRs are created or updated.

andrewdonheiser committed:
ci: Run push tests on all commits

Update workflow to trigger on all pushes (not just main branch) so tests
run on every commit to feature branches, providing immediate feedback.

andrewdonheiser committed:
fix: Resolve all ruff linting issues in test files

- Remove unused imports (nbformat, sys, Path, tempfile)
- Remove unnecessary 'r' mode arguments from open() calls (UP015)
- Fix unused loop variables by prefixing with underscore (B007)
- Remove unused local variable has_imports (F841)
- Fix import sorting in test_utils.py (I001, E402)
- Remove trailing whitespace from blank lines (W293)

All ruff linting errors resolved.

andrewdonheiser committed:
fix: Add noqa comment for intentional E402 in test_utils.py

The knowledge_utils import must come after sys.path modification,
which is intentional. Add noqa comment to suppress E402 error.

andrewdonheiser committed:
style: Fix ruff formatting issues in test files

- Add blank lines after docstrings in all test files
- Break long lines to comply with line length limits
- Reformat multi-line assert statements
- Reformat long conditional expressions
- Remove trailing blank lines

All formatting issues resolved to pass ruff format checks.

andrewdonheiser committed:
docs: Fix markdown linting errors in AGENTS.md and README.md

- Add blank lines around fenced code blocks (MD031)
- Add blank lines around lists (MD032)
- Remove multiple consecutive blank lines (MD012)

All markdown linting errors resolved.

kami619 commented on .github/workflows/knowledge-tuning-push-tests.yml:
```diff
@@ -0,0 +1,32 @@
+name: Knowledge Tuning Push Tests
```
did you mean `PR Merge` tests ? it might be worth to rename the file and action to something everyone would understand easily and reduce the ambiguity.

andrewdonheiser replied:
these are meant to be run when pushed to remote. I wasn't sure exactly how they wanted to run these.

kami619 commented on tests/knowledge-tuning/README.md:
```diff
@@ -0,0 +1,98 @@
+# Knowledge Tuning Push Tests
```
given this is a new entity, might be worth to add a section that details about the ground rules on Contributions and point to any PR templates that they might not be aware of during the development phase.

andrewdonheiser replied:
👍

andrewdonheiser committed:
tightened agent rules about folders it can change files in

andrewdonheiser committed:
tightened agent rules

andrewdonheiser committed:
linting fix

tarun-etikala commented on tests/format_checks/test_notebook_structure.py:
This is a nice thorough tests for notebook structure. Can we have it run for all the notebooks in examples/ instead of just knowledge tuning. 
And some of the expectations like "First cell should typically be markdown" can be documented in [CONTRIBUTING.md](https://github.com/red-hat-data-services/red-hat-ai-examples/blob/main/CONTRIBUTING.md)

andrewdonheiser replied:
sounds good. I will expand the scope and update CONTRIBUTING.md

andrewdonheiser committed:
# Notebook Format & Knowledge-Tuning Push Tests

## Summary

- Split the original push-test suite into two focused packages: repo-wide notebook format checks (`tests/format_checks/`) and knowledge-tuning specific tests (`tests/knowledge-tuning/`).
- Introduced shared infrastructure under `tests/` (common `pyproject.toml`, session fixtures, shared `test-results/` root, and documentation) with `AGENTS.md` as the single source of truth.
- Updated the GitHub Actions workflow to install shared deps once and run both runners on every push/pull request.
- Refreshed documentation (`tests/AGENTS.md`, `tests/format_checks/README.md`, `tests/knowledge-tuning/README.md`, `tests/README.md`, `CONTRIBUTING.md`, `PR_MESSAGE.md`) to describe the new structure, artifact locations, and expectations; fixed markdownlint spacing issues.

## Test Infrastructure

### Suites
- **Repo-wide Format Checks (`tests/format_checks/`)**
  - `test_notebook_structure.py`: Metadata, outputs, ordering, section enforcement across every notebook in the repo.
  - `test_notebook_imports.py`: Static AST/tokenize validation of all notebook imports.
  - `test_environment.py`: Ensures every `pyproject.toml` parses, declares dependencies correctly, and uses modern Python versions.
  - `test_venv_build.py`: `pip install --dry-run` + GPU-aware skips for dependency resolution.
  - `format_checks.py`: Runner that enables parallel pytest execution and publishes HTML/JUnit artifacts.
- **Knowledge-Tuning (`tests/knowledge-tuning/`)**
  - `test_utils.py`: Mock-driven coverage for `knowledge_utils.py`.
  - `run_knowledge_tuning_tests.py`: Updated runner that mirrors the reporting workflow used by format checks.
  - `mocks/`: Existing deterministic mocks retained for transformers/tokenizers.

### Shared Features
- `tests/conftest.py`: Provides a repo root fixture for every suite.
- `tests/pyproject.toml`: Central dependency list for both runners (`pip install -e tests`).
- `tests/AGENTS.md`: Updated to describe both suites, setup commands, mocking requirements, CI scope, and markdownlint-compliant formatting.
- `tests/test-results/`: Shared artifact root with per-suite subfolders; both runners generate `{junit.xml,report.html}` there for CI ingestion or local review.

## Files Added / Updated

- **New Structure**
- `tests/format_checks/` (conftest, tests, README, runner)
  - `tests/knowledge-tuning/README.md` (new human-focused guide)
  - `tests/conftest.py`, `tests/pyproject.toml`, `tests/AGENTS.md`
- **Renamed**
  - `tests/knowledge-tuning/run_push_tests.py` → `run_knowledge_tuning_tests.py`
  - `.github/workflows/knowledge-tuning-push-tests.yml` → `.github/workflows/notebook-tests.yml`
- **Documentation**
  - `CONTRIBUTING.md`: New section describing mandatory notebook test runs.
  - `PR_MESSAGE.md`: Updated to capture the new scope for this PR.

## Documentation Highlights

- `tests/AGENTS.md`: Now documents setup, running instructions, mocking policy, CI expectations, and complies with markdownlint spacing rules.
- `tests/format_checks/README.md`: Explains repo-wide coverage, commands, artifact locations, and follows markdownlint spacing rules.
- `tests/knowledge-tuning/README.md`: Clarifies that knowledge-specific tests remain focused on mocked utility coverage.
- `tests/README.md`: Documents the shared structure, dependency install flow, and the centralized `tests/test-results/<suite>/` artifacts.
- `CONTRIBUTING.md`: Instructs contributors to run `python tests/format_checks/format_checks.py` and `python tests/knowledge-tuning/run_knowledge_tuning_tests.py` before pushing.

## CI / Automation

- `.github/workflows/notebook-tests.yml`: Installs shared deps (`cd tests && pip install -e .`) and runs both runners on every push and PR targeting `main`.
- Workflow timeout bumped to 15 minutes, though suites remain sub-minute locally thanks to pytest-xdist.

## Testing

```bash
cd tests
pip install -e .

python tests/format_checks/format_checks.py
python tests/knowledge-tuning/run_knowledge_tuning_tests.py
```

Each runner reports to `tests/test-results/<suite>/` (HTML + JUnit). Existing notebook cleanliness issues are still surfaced by the format suite (expected failures until notebooks are scrubbed).

## Benefits

1. **Repo-wide Coverage** – Every notebook and `pyproject.toml` now gets validated on push.
2. **Reusable Infra** – Shared `pyproject.toml`, fixtures, and docs simplify future expansion.
3. **Faster CI** – Single dependency installation for both suites plus parallel pytest workers.
4. **Clear Guidance** – AGENTS + README + CONTRIBUTING all describe the new expectations for contributors and agents.

tarun-etikala commented:
Opened a new [pull request](https://github.com/red-hat-data-services/red-hat-ai-examples/pull/24) with rebase and test cleanup. Closing this one.
